doc/03_research/03_planning/06_state_machine_design.md

```diff
@@ -0,0 +1,278 @@
+
+### Brake
+
+Transition:
+
+* From `KEEP` if `new target_speed` is smaller than current `target_speed` or an `obstacle` or the `leading_vehicle` is in braking distance.
+* From `BRAKE` if `new target_speed` is smaller than current `target_speed` or an `obstacle` or the `leading_vehicle` is in braking distance.
+
+Set a new target speed and change back to `KEEP` state afterwards.
+
+## Lane change state machine
+
+![img.png](../../00_assets/Lane_Change_SM.png)
+
+Transition:
+
+* From `driving state machine` by `lane_change_requested`
+
+This state machine completes the change of a lane. This is triggered from the super state machine and can have multiple triggers. Those include:
+
+* Join highway
+* Leave highway
+* RoadOption:
+  * CHANGELANELEFT
+  * CHANGELANERIGHT
+  * KEEPLANE
+* avoid obstacle(doors, static objects)
+* give way to emergency vehicle
+* overtake slow moving vehicle
+* leave a parking bay
+
+### DECIDE_LANE_CHANGE
+
+Transition:
+
+* From super state machine by above triggers
+
+From the super state machine the transition to change the lane is given by one of the above triggers. This state decides to which lane should be changed dependent on the trigger.
+It takes into account if there are lanes to the left and/or right and if the lane change is requested by a roadOption command.
+
+### CHANGE_LANE_LEFT
+
+Transition:
+
+* From `DECIDE_LANE_CHANGE` by `RoadOption.CHANGELANELEFT` or `obstacle_in_lane` or `leader_vehicle_speed < LEADERTHRESHOLD`
+
+This state performs a lane change to the lane on the left.
+
+If the lane is for oncoming traffic, it should be checked if there is enough time to complete two lane changes plus a buffer for the length of an obstacle without an accident.
+
+If the lane is for traffic flowing in the same direction, It should be checked if the lane is free and if there is no faster moving vehicles coming from behind.
+
+If an obstacle or a slow leading vehicle are the reasons for the lane change, to join the same lane after the obstacle is passed, the state changes to `CHANGE_LANE_RIGHT`.
+
+### CHANGE_LANE_RIGHT
+
+Transition:
+
+* From `DECIDE_LANE_CHANGE` by `RoadOption.CHANGELANERIGHT` or `emergency_vehicle_in_front`
+* From `CHANGE_LANE_LEFT` by `passing_obstacle` or `slow_leading_vehicle`
+
+For changing to the right lane it is assumed, that the traffic in this lane flows in the driving direction of the ego vehicle.
+
+The lane change should be performed if the lane is free and there are no fast moving cars behind us that would crash with the ego vehicle.
+
+## Intersection state machine
+
+![img.png](../../00_assets/Intersection_SM.png)
+
+Transition:
+
+* From `driving state machine` by `intersection_detected`
+
+This state machine handles the passing of an intersection.
+
+### APPROACH_INTERSECTION
+
+Slow down in the approach of an intersection.
+
+If the intersection is `unsignalized` and `clear` change to the `ENTER_INTERSECTION` state.
```

Comment:
Was passiert, wenn die intersection `unsignalized`, aber _nicht_ `clear` ist?

Reply:
Dann bleibt die State machine in dem Zustand

Reply:
Das würde bedeuten, dass wir weiter auf die Kreuzung zufahren. Wird die Kreuzung dann weiter nicht frei, wird auch nicht in den STOP_TRAFFIC_SM gewechselt, da es kein Stoppschild und keine Ampel gibt. Somit werden wir früher oder später auf in die Kreuzung einfahren, obwohl sie nicht frei ist, wenn ich das richtig sehe.

Reply:
Wenn die Kreuzung unbeschildert ist, wird nie in die STOP_TRAFFIC_SM gewechselt, sie ist für diesen Fall nicht ausgelegt, da weder Stopschilder noch Ampeln vorhanden sind. Das Ziel von dem APPROACH zustand ist die Kreuzung anzufahen und notfalls vor der Kreuzung zu warten. Das geht aber aus der Grafik nicht hervor.

Reply:
Ich würde im Zweifelsfall immer vor der Kreuzung bremsen, wenn sie nicht frei ist. Es kann ja auch passieren, dass wir die Schilder oder Ampeln einfach nur nicht erkannt haben. Auf deutschen Straßen hat man auch noch Rechts-vor-Links, da gibt es auch keine Schilder

Reply:
Im Carla Leaderboard, gilt aber explizit, dass bei unsignalisierten Kreuzungen, der Vorfahrt hat, der als erstes in der Kreuzung ist. Und die Gescwindigkeit wird beim Approach pauschal verringert